scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/EcdhPskServerKeyExchange.java

```diff
@@ -0,0 +1,247 @@
+import java.security.spec.ECParameterSpec;
+import java.security.spec.ECPoint;
+import java.security.spec.ECPublicKeySpec;
+import java.util.Arrays;
+
+import org.eclipse.californium.elements.util.DatagramReader;
+import org.eclipse.californium.elements.util.DatagramWriter;
+import org.eclipse.californium.elements.util.StringUtil;
+import org.eclipse.californium.scandium.dtls.AlertMessage.AlertDescription;
+import org.eclipse.californium.scandium.dtls.AlertMessage.AlertLevel;
+import org.eclipse.californium.scandium.dtls.cipher.ECDHECryptography;
+import org.eclipse.californium.scandium.dtls.cipher.ECDHECryptography.SupportedGroup;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Generates server ephemeral ECDH keys for Dtls PSK mode.
+ * <p>
+ * Server must send the {@code ServerKeyExchange} message even if the PSK identity hint is not provided.
+ * 
+ * @see <a href="https://tools.ietf.org/html/rfc5489#section-2">RFC 5489</a>
+ *
+ */
+public final class EcdhPskServerKeyExchange extends ServerKeyExchange{
+	
+	private static final Logger LOGGER = LoggerFactory.getLogger(EcdhPskServerKeyExchange.class.getCanonicalName());
+	
+	private static final String MSG_UNKNOWN_CURVE_TYPE = "Unknown curve type [{}]";
+	private static final int CURVE_TYPE_BITS = 8;
+	private static final int NAMED_CURVE_BITS = 16;
+	private static final int PUBLIC_LENGTH_BITS = 8;
+	
+	/**
+	 * The algorithm name to generate elliptic curve keypairs. See also <a href=
+	 * "http://docs.oracle.com/javase/7/docs/technotes/guides/security/StandardNames.html#KeyPairGenerator"
+	 * >KeyPairGenerator Algorithms</a>.
+	 */
+	private static final String KEYPAIR_GENERATOR_INSTANCE = "EC";
+
+	/** The ECCurveType */
+	
+	/** parameters are conveyed verbosely; underlying finite field is a prime 
+	 * field
+	*/ 
+	private static final int EXPLICIT_PRIME = 1;
+	
+	/** parameters are conveyed verbosely; underlying finite field is a 
+	 * characteristic-2 field
+	*/ 
+	private static final int EXPLICIT_CHAR2 = 2;
+	
+	/** a named curve is used */
+	private static final int NAMED_CURVE = 3;
+
+	/** ephemeral public key */
+	private ECPublicKey publicKey = null;
+
+	private ECPoint point = null;
+	private byte[] pointEncoded = null;
+
+	private final int curveId;
+	
+	// TODO right now only named curve is supported
+	private int curveType = NAMED_CURVE;
+	
+	/**
+	 * Creates a new key exchange message with ServerDHParams.
+	 * See <a href="http://tools.ietf.org/html/rfc4279#section-3">RFC 4279</a>:
+	 * 
+	 * @param ecdhe - {@code ECDHECryptography}
+	 * @param clientRandom - nonce
+	 * @param serverRandom - nonce
+	 * @param namedCurveId - ec curve used 
+	 * @param peerAddress
+	 * @throws GeneralSecurityException
+	 * @throws NullPointerException if any of arguments ecdhe, 
+	 * clientRandom, serverRandom are {@code null}
+	 */
+	public EcdhPskServerKeyExchange(ECDHECryptography ecdhe, Random clientRandom, Random serverRandom, 
```

Comment:
should allow specification of an identity hint

Reply:
added identity hint in the constructor